File: cli/secret_sharing.py
import sys
import base64
import hashlib
import json
import requests
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
import os

# Encrypt secret with AES-GCM (no padding, compatible with Web Crypto API)
def encrypt_secret(secret, password):
    backend = default_backend()
    salt = b'salt'  # Same salt as in JavaScript
    key = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 100000, dklen=32)
    iv = os.urandom(12)  # 12-byte IV for GCM
    cipher = Cipher(algorithms.AES(key), modes.GCM(iv), backend=backend)
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(secret.encode()) + encryptor.finalize()
    # Combine ciphertext and tag (16 bytes) as Web Crypto API does
    encrypted = ciphertext + encryptor.tag[:16]
    return base64.b64encode(iv).decode(), base64.b64encode(encrypted).decode()

# Decrypt secret with AES-GCM (no padding, compatible with Web Crypto API)
def decrypt_secret(encrypted, iv, password):
    backend = default_backend()
    salt = b'salt'  # Same salt as in JavaScript
    key = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 100000, dklen=32)
    iv_bytes = base64.b64decode(iv)
    encrypted_bytes = base64.b64decode(encrypted)
    # Split ciphertext and tag (last 16 bytes are the tag)
    ciphertext = encrypted_bytes[:-16]
    tag = encrypted_bytes[-16:]
    cipher = Cipher(algorithms.AES(key), modes.GCM(iv_bytes, tag), backend=backend)
    decryptor = cipher.decryptor()
    plaintext = decryptor.update(ciphertext) + decryptor.finalize()
    return plaintext.decode()

# Post secret to API
def post_secret(url, master_password, secret):
    iv, encrypted = encrypt_secret(secret, master_password)
    payload = {
        "encrypted": encrypted,
        "iv": iv,
        "expiry": 86400,  # 1 day default
        "maxAttempts": 3,
        "isE2EE": True
    }
    response = requests.post(f"{url}/api/create", json=payload)
    data = response.json()
    if data.get("success"):
        link = f"{url}/retrieve.html?code={data['code']}"
        print(f"Secret created!\nLink: {link}\nOTP: {data['otp']}")
    else:
        print(f"Error: {data.get('message')}")

# Retrieve and decrypt secret
def retrieve_secret(url, master_password, code, otp):
    response = requests.get(f"{url}/api/retrieve?code={code}&otp={otp}")
    data = response.json()
    if data.get("success"):
        decrypted = decrypt_secret(data["encrypted"], data["iv"], master_password)
        print(f"Decrypted Secret: {decrypted}")
    else:
        print(f"Error: {data.get('message')}")

# Main script
if __name__ == "__main__":
    url = sys.argv[1] if len(sys.argv) > 1 else input("Enter API URL (e.g., https://your-app.pages.dev): ")
    master_password = sys.argv[2] if len(sys.argv) > 2 else input("Enter master password: ")
    action = input("Choose action (1 = Post Secret, 2 = Retrieve Secret): ")

    if action == "1":
        secret = sys.argv[3] if len(sys.argv) > 3 else input("Enter secret: ")
        post_secret(url, master_password, secret)
    elif action == "2":
        code = sys.argv[3] if len(sys.argv) > 3 else input("Enter secret code: ")
        otp = sys.argv[4] if len(sys.argv) > 4 else input("Enter OTP: ")
        retrieve_secret(url, master_password, code, otp)
    else:
        print("Invalid action. Use 1 or 2.")

File: cli/secret_sharing.js
const crypto = require('crypto');

// Encrypt secret with AES-GCM (no padding, tag appended)
async function encryptSecret(secret, password) {
  const salt = Buffer.from('salt'); // Same salt as in web app
  const key = crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha256');
  const iv = crypto.randomBytes(12); // 12-byte IV
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
  const encrypted = Buffer.concat([cipher.update(secret, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag(); // 16-byte tag
  const combined = Buffer.concat([encrypted, tag]); // Append tag to ciphertext
  return { iv: iv.toString('base64'), data: combined.toString('base64') };
}

// Decrypt secret with AES-GCM (extract tag from combined data)
async function decryptSecret(encrypted, iv, password) {
  const salt = Buffer.from('salt'); // Same salt as in web app
  const key = crypto.pbkdf2Sync(password, salt, 100000, 32, 'sha256');
  const encryptedBytes = Buffer.from(encrypted, 'base64');
  const ciphertext = encryptedBytes.slice(0, -16); // All but last 16 bytes
  const tag = encryptedBytes.slice(-16); // Last 16 bytes are the tag
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, Buffer.from(iv, 'base64'));
  decipher.setAuthTag(tag);
  const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  return decrypted.toString('utf8');
}

// Post secret to API
async function postSecret(url, masterPassword, secret) {
  const { iv, data } = await encryptSecret(secret, masterPassword);
  const payload = {
    encrypted: data,
    iv: iv,
    expiry: 86400, // 1 day default
    maxAttempts: 3,
    isE2EE: true
  };
  const response = await fetch(`${url}/api/create`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  const result = await response.json();
  if (result.success) {
    const link = `${url}/retrieve.html?code=${result.code}`;
    console.log(`Secret created!\nLink: ${link}\nOTP: ${result.otp}`);
  } else {
    console.log(`Error: ${result.message}`);
  }
}

// Retrieve and decrypt secret
async function retrieveSecret(url, masterPassword, code, otp) {
  const response = await fetch(`${url}/api/retrieve?code=${code}&otp=${otp}`);
  const data = await response.json();
  if (data.success) {
    const decrypted = await decryptSecret(data.encrypted, data.iv, masterPassword);
    console.log(`Decrypted Secret: ${decrypted}`);
  } else {
    console.log(`Error: ${data.message}`);
  }
}

// Main script
const [,, url, masterPassword, ...rest] = process.argv;
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

(async () => {
  const finalUrl = url || await new Promise(resolve => readline.question("Enter API URL (e.g., https://your-app.pages.dev): ", resolve));
  const finalPassword = masterPassword || await new Promise(resolve => readline.question("Enter master password: ", resolve));
  const action = await new Promise(resolve => readline.question("Choose action (1 = Post Secret, 2 = Retrieve Secret): ", resolve));

  if (action === "1") {
    const secret = rest[0] || await new Promise(resolve => readline.question("Enter secret: ", resolve));
    await postSecret(finalUrl, finalPassword, secret);
  } else if (action === "2") {
    const code = rest[0] || await new Promise(resolve => readline.question("Enter secret code: ", resolve));
    const otp = rest[1] || await new Promise(resolve => readline.question("Enter OTP: ", resolve));
    await retrieveSecret(finalUrl, finalPassword, code, otp);
  } else {
    console.log("Invalid action. Use 1 or 2.");
  }
  readline.close();
})();